azurecompute/src/main/java/org/jclouds/azurecompute/compute/extensions/AzureComputeSecurityGroupExtension.java

```diff
@@ -0,0 +1,460 @@
+           priority, IpPermission ipPermission) {
+
+      String protocol = ipPermission.getIpProtocol().name();
+      String destinationPortRange;
+      if (ipPermission.getFromPort() != ipPermission.getToPort()) {
+         destinationPortRange = String.format("%s-%s", ipPermission.getFromPort(), ipPermission.getToPort());
+      } else {
+         destinationPortRange = String.valueOf(ipPermission.getToPort());
+      }
+      final String destinationAddressPrefix =
+              ipPermission.getCidrBlocks().isEmpty() || Iterables.get(ipPermission.getCidrBlocks(), 0).equals("0.0.0.0/0") ?
+                      "*" : Iterables.get(ipPermission.getCidrBlocks(), 0);
+      String setRuleToNSGRequestId = api.getNetworkSecurityGroupApi().setRule(networkSecurityGroupId, ruleName,
+              Rule.create(ruleName, // name
+                      "Inbound", // type
+                      String.valueOf(priority), // priority
+                      "Allow", // action
+                      "INTERNET", // sourceAddressPrefix
+                      "*", // sourcePortRange
+                      destinationAddressPrefix, // destinationAddressPrefix
+                      destinationPortRange, // destinationPortRange
+                      protocol,  // protocol
+                      "Active", // state
+                      true // isDefault
+              ));
+      if (!operationSucceeded.apply(setRuleToNSGRequestId)) {
+         // TODO
+      }
+   }
+
+   private void removeRuleFromNetworkSecurityGroup(String id, String ruleName) {
+      String setRuleToNSGRequestId = api.getNetworkSecurityGroupApi().deleteRule(id, ruleName);
+      if (!operationSucceeded.apply(setRuleToNSGRequestId)) {
+         // TODO
+      }
+   }
+
+   private Role.ConfigurationSet.InputEndpoint createInputEndpoint(String ruleName, String protocol, String address, int port) {
+      return Role.ConfigurationSet.InputEndpoint.create(
+              ruleName,
+              protocol,
+              port,
+              port,
+              address,
+              false, // enabledDirectServerReturn
+              null, // loadBalancerName
+              null, // loadBalancerProbe
+              null //idleTimeoutInMinutes
+      );
+   }
+
+   public void waitForOperationCompletion(String operationId, String operationDescription) {
+      if (!operationSucceeded.apply(operationId)) {
+         final String warnMessage = format("%s) has not been completed within %sms.", operationDescription, Long.toString(operationTimeout));
+         logger.warn(warnMessage);
+         String illegalStateExceptionMessage = format("%s. Please, try by increasing `jclouds.azure.operation-timeout` and try again", warnMessage);
+         throw new IllegalStateException(illegalStateExceptionMessage);
+      } else {
+         logger.info("%s (id: %s) terminated successfully", operationDescription, operationId);
+      }
+   }
```

Comment:
This method is already present in the compute service adapter. Extract it somewhere so it can be reused.